feat(Auth): Adding documentation for TOTP support (#5690)
aws-amplify/docs: src/fragments/lib/auth/common/mfa/flows.mdx

The Auth category supports Multi-factor Authentication (MFA) for user sign-in flows. MFA is an extra layer of security used to make sure that people trying to gain access to an online account are who they say they are. It requires users to provide additional information to verify their identity. The category supports the following MFA methods:
* SMS
* TOTP (Time-based One-time Password)

## Set Up Backend Resources

<Callout warning>

When using Amplify CLI to set up backend resources, the following options are only available when starting a new project (via `amplify add auth`). You will not have access to these settings after creation (via `amplify update`).

- Required MFA (i.e. Setting MFA to "ON")

</Callout>

<Callout>
Note: If you create or update an SMS MFA configuration for your Cognito user pool, the Cognito service will send a test SMS message to an internal number in order to verify your configuration. You will be charged for these test messages by Amazon SNS. 

> Refer [SMS sandbox](/lib/auth/sms_flows#sandbox-mode).

Reviewer: ```suggestion
Refer to [SMS sandbox](/lib/auth/sms_flows#sandbox-mode).
```

For information about Amazon SNS pricing, see [Worldwide SMS Pricing](https://aws.amazon.com/sns/sms-pricing/). 
</Callout>

<BlockSwitcher>

<Block name="New Project">

Run `amplify add auth` to create a new Cognito Auth resource, and follow the prompts below depending on how you want to integrate MFA into your flow.

Turning MFA "ON" will make it required for all users, while "Optional" will make it available to enable on a per-user basis.

import all2 from "/src/fragments/lib/auth/common/mfa/add_mfa.mdx";

<Fragments fragments={{all: all2}} />

</Block>

<Block name="Existing Auth">

Run `amplify update auth` and follow the prompts as guided below.

import all5 from "/src/fragments/lib/auth/common/mfa/update_mfa.mdx";

<Fragments fragments={{all: all5}} />

</Block>

</BlockSwitcher>

## Multi-factor authentication with SMS

<Callout>

> <b>AWS Cognito Limitation</b>

Reviewer: This doesn't seem to belong here. Maybe put in the `Setting preferred MFA` section below.

Author: That is correct. updated.

Author: Refer
https://github.com/aws-amplify/docs/pull/5690#discussion_r1297802883

The `fetchMFAPreference` API will only return enabled and preferred MFA methods, if the preferences were set using the `updateMFAPreference` API. If the preferences were not set using the `updateMFAPreference` API, the MFA preferences will return `nil` for enabled and preferred MFA methods. 

Therefore it is recommended to use the `updateMFAPreference` API to set the MFA preferences for the user, if the user decides to set up any MFA methods. An example use case would be that when a user successfully sets up TOTP using the `setupTOTP` and `verifyTOTPSetup` API, `updateMFAPreference` should be invoked to set the MFA preferences for the user i.e. marking TOTP as enabled.

</Callout>

### Adding phone number during Sign Up

> Sign up users normally with the chosen `Username` type and password. `"phone_number"` will be required in the `userAttributes` if SMS MFA is ON, or manually enabled for the user

Reviewer: There's some nuance here, though. If the username attribute is `phone_number` then it must not be included in `userAttributes` IIRC

Author: Got it. Updated the wording to include a note.

Verification of user accounts is done via the `confirmSignUp` method with the OTP sent to their phone or email.

import flutter7 from "/src/fragments/lib/auth/flutter/sms/confirm_sign_up.mdx";

<Fragments fragments={{flutter: flutter7}} />

import iosConfirmSignUp from "/src/fragments/lib/auth/ios/sms/confirm_sign_up.mdx";

<Fragments fragments={{ios: iosConfirmSignUp}} />

### Handling SMS MFA challenge during Sign In

Sign in users normally with the chosen `Username` type and password.

import flutter8 from "/src/fragments/lib/auth/flutter/sms/sign_in.mdx";

<Fragments fragments={{flutter: flutter8}} />

import iosSignInTOTP from "/src/fragments/lib/auth/ios/sms/sign_in.mdx";

<Fragments fragments={{ios: iosSignInTOTP}} />

If MFA is **ON** or enabled for the user, you must call `confirmSignIn` with the OTP sent to their phone.

import flutter9 from "/src/fragments/lib/auth/flutter/sms/confirm_sign_in.mdx";

<Fragments fragments={{flutter: flutter9}} />

import iosConfirmSignInTOTP from "/src/fragments/lib/auth/ios/sms/confirm_sign_in.mdx";

<Fragments fragments={{ios: iosConfirmSignInTOTP}} />

## Multi-factor authentication with TOTP

<Callout>

> <b>AWS Cognito Limitation</b>

Reviewer: Same comment as above

Author: Refere https://github.com/aws-amplify/docs/pull/5690#discussion_r1297802883

The `fetchMFAPreference` API will only return enabled and preferred MFA methods, if the preferences were set using the `updateMFAPreference` API. If the preferences were not set using the `updateMFAPreference` API, the MFA preferences will return `nil` for enabled and preferred MFA methods. 

Therefore it is recommended to use the `updateMFAPreference` API to set the MFA preferences for the user, if the user decides to set up any MFA methods. An example use case would be that when a user successfully sets up TOTP using the `setupTOTP` and `verifyTOTPSetup` API, `updateMFAPreference` should be invoked to set the MFA preferences for the user i.e. marking TOTP as enabled.

</Callout>

### Handling TOTP Setup during sign in

A `signIn` call will return `continueSignInWithTOTPSetup` as a next step when the following set of conditions are met:

- MFA is marked as **Required** in Cognito User Pool.
- TOTP is marked as an enabled MFA method in Cognito user pool
- User does not have any other MFA method set up.

> `continueSignInWithTOTPSetup` step signifies that the user must set up TOTP before they can sign in. The step returns an associated value of type `TOTPSetupDetails` which would be used for generating TOTP. `TOTPSetupDetails` provides a helper method called `getSetupURI` that can be used to generate a URI, which can be used by native password managers for TOTP association. For example. if the URI is used on Apple platforms, it will trigger the platform's native password manager to associate TOTP with the account. For more advanced use cases, `TOTPSetupDetails` also contains the `sharedSecret` that will be used to either generate a QR code or can be manually entered into an Authenticator app. 

Reviewer: ```suggestion
The `continueSignInWithTOTPSetup` step signifies that the user must set up TOTP before they can sign in. The step returns an associated value of type `TOTPSetupDetails` which must be used to configure an Authenticator app like Microsoft Authenticator or Google Authenticator. `TOTPSetupDetails` provides a helper method called `getSetupURI` which generates a URI that can be used, for example, in a button to open the user's installed Authenticator app. For more advanced use cases, `TOTPSetupDetails` also contains a `sharedSecret` which can be used to either generate a QR code or be manually entered into an Authenticator app. 
```

> Once the authenticator app is set up, the user can generate a TOTP code and provide it to the library to complete the sign in process.

Reviewer: ```suggestion
Once the Authenticator app is set up, the user must generate a TOTP code and provide it to the library to complete the sign in process.
```

import iosSignIn from "/src/fragments/lib/auth/ios/totp/sign_in.mdx";

<Fragments fragments={{ios: iosSignIn}} />

The TOTP code can be obtained from the user via a text field or any other means. Once the user provides the TOTP code, call `confirmSignIn` with the TOTP code as the `challengeResponse` parameter.

import iosConfirmSignIn from "/src/fragments/lib/auth/ios/totp/confirm_sign_in.mdx";

<Fragments fragments={{ios: iosConfirmSignIn}} />

### Enabling TOTP after a user is signed in

TOTP MFA can be set up after a user has signed in. This can be done when the following conditions are met:

> - MFA is marked as **Optional** or **Required** in Cognito User Pool.

Reviewer: ```suggestion
- MFA is marked as **Optional** or **Required** in the Cognito User Pool
```

> - TOTP is marked as an enabled MFA method in Cognito user pool

Reviewer: ```suggestion
- TOTP is enabled in the Cognito User Pool
```

> TOTP setup can be achieved by calling `setUpTOTP` and `verifyTOTPSetup` APIs that are available in the `Auth` category.

Reviewer: ```suggestion
TOTP can be set up by calling the `setUpTOTP` and `verifyTOTPSetup` APIs in the `Auth` category.
```

> Invoke the `setUpTOTP` API to generate a `TOTPSetupDetails` object that contains the `sharedSecret` which will be used to either to generate a QR code or can be manually entered into an Authenticator app.

Reviewer: I would more-or-less copy the spiel from above and mentioning `getSetupUri`. The inclusion of shared secret should be an afterthought in the discussion IMO.

**`setUpTOTP` API**

import iosSetUpTOTP from "/src/fragments/lib/auth/ios/totp/set_up_totp.mdx";

<Fragments fragments={{ios: iosSetUpTOTP}} />

> Once the authenticator app is set up, the user can get a TOTP code that is generated by the Authenticator app and can provide it to the library to using the `verifyTOTPSetup` API to complete the TOTP setup process.

Reviewer: ```suggestion
Once the Authenticator app is set up, the user must generate a TOTP code and provide it to the library. Pass the code to `verifyTOTPSetup` to complete the TOTP setup process.
```

**`verifyTOTPSetup` API**

import iosVerifyTOTPSetup from "/src/fragments/lib/auth/ios/totp/verify_totp_setup.mdx";

<Fragments fragments={{ios: iosVerifyTOTPSetup}} />

### Recovering from a lost TOTP device

> If a user loses access to their TOTP device, they would need to contact an administrator to help get access to their account. Based on the Cognito User Pool configuration, the administrator can use the [AdminSetUserMFAPreference](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_AdminSetUserMFAPreference.html) to either change the MFA preference to a different MFA method or to disable MFA for the user.

Reviewer: I would think this would want to be called out in the warning. This is a huge deal IMO.

Author: Done.

> <Callout warning>

Author: Done.

In a scenario where MFA is marked as **Required** in Cognito User Pool and another MFA method is not set up, the administrator would need to first initiate an [AdminUpdateUserAttributes](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_AdminUpdateUserAttributes.html) call and update the user’s phone number attribute. Once this is complete, the administrator can continue changing the MFA preference to SMS as suggested above.

</Callout>

> ## Setting up user preferred MFA option

Reviewer: ```suggestion
## Setting a user's preferred MFA option
```

> You can use Time-based One-Time Password (TOTP) for multi-factor authentication (MFA) in your web or mobile applications. Amplify's Auth category includes support for handling TOTP setup and codes, offering an integrated solution for enhanced security for your users. Authenticator apps generate the temporary, one-time passwords. These apps, such as Google Authenticator, Microsoft Authenticator, have the TOTP algorithm built-in and work by using a shared secret key and the current time to generate a short-lived password.

Reviewer: Is this paragraph needed here? Probably only if it's going to be its own page, but TOTP is well-established by this point on this page.

Author: I agree. Moved some of it to the TOTP section.

<Callout>

> The APIs for fetching and updating MFA preferences are only available on the AWS Cognito Auth plugin and not on the Auth category. 

Author: May be I should mention as Escape Hatch API's.

> The APIs for fetching and updating MFA preferences are only available on the AWS Cognito Auth plugin and not on the Auth category. 

Author: Updated the wording a little bit, but there is no existing reference of it anywhere in the doc.

</Callout>

### Fetch the current user's MFA preferences

Invoke the following API to get the MFA preference for the current user.

import iosFetchMFAPreference from "/src/fragments/lib/auth/ios/mfa_preference/10_fetch_mfa_preference.mdx";

<Fragments fragments={{ios: iosFetchMFAPreference}} />

### Update the current user's MFA preferences

Invoke the following API to update the MFA preference for the current user.

<Callout warning>

Only one MFA method can be marked as preferred at a time. If the user has multiple MFA methods enabled and tries to mark more than one MFA method as preferred, the API will throw an error.

</Callout>

import iosUpdateMFAPreferences from "/src/fragments/lib/auth/ios/mfa_preference/20_update_mfa_preference.mdx";

<Fragments fragments={{ios: iosUpdateMFAPreferences}} />

If multiple MFA methods are enabled for the user, the `signIn` API will return `continueSignInWithMFASelection` as the next step in the auth flow. During this scenario, the user should be prompted to select the MFA method they want to use to sign in.

> Refer to the [Continue signin with MFA Selection](/lib/auth/signin_next_steps#continue-signin-with-mfa-selection) section for more details.

Author: Updated.

> The Auth category supports Multi-factor Authentication (MFA) for user sign-in flows. MFA is an extra layer of security used to make sure that people trying to gain access to an online account are who they say they are. It requires users to provide additional information to verify their identity. The category supports the following MFA methods:

Reviewer 2: Small nits:

The Auth category supports Multi-factor Authentication (MFA) for user sign-in flows. MFA is an extra layer of security used to make sure that users trying to gain access to an account are who they say they are. It requires users to provide additional information to verify their identity. The category supports the following MFA methods:

> * SMS

Reviewer 2: these should anchor link to each section

* TOTP (Time-based One-time Password)

> ## Set Up Backend Resources

Reviewer 2: Right under this title:

Below are the steps you can use to set up MFA using SMS or TOTP with the Amplify CLI. The Amplify libraries are designed to work with MFA even if you have set up your Amazon Cognito resources separately.

import all2 from "/src/fragments/lib/auth/common/mfa/add_mfa.mdx";

<Fragments fragments={{all: all2}} />

</Block>

<Block name="Existing Auth">

Run `amplify update auth` and follow the prompts as guided below.

import all5 from "/src/fragments/lib/auth/common/mfa/update_mfa.mdx";

<Fragments fragments={{all: all5}} />

</Block>

</BlockSwitcher>

## Multi-factor authentication with SMS

<Callout>

<b>AWS Cognito Limitation</b>

The `fetchMFAPreference` API will only return enabled and preferred MFA methods, if the preferences were set using the `updateMFAPreference` API. If the preferences were not set using the `updateMFAPreference` API, the MFA preferences will return `nil` for enabled and preferred MFA methods. 

Therefore it is recommended to use the `updateMFAPreference` API to set the MFA preferences for the user, if the user decides to set up any MFA methods. An example use case would be that when a user successfully sets up TOTP using the `setupTOTP` and `verifyTOTPSetup` API, `updateMFAPreference` should be invoked to set the MFA preferences for the user i.e. marking TOTP as enabled.

</Callout>

> ### Adding phone number during Sign Up

Reviewer 2: Enabling SMS for MFA during Sign Up

> Sign up users normally with the chosen `Username` type and password. `"phone_number"` will be required in the `userAttributes` if SMS MFA is ON, or manually enabled for the user. Note that if the username attribute is the `phone_number`, then adding a separate attribute for phone number is not required during sign up.

Reviewer 2: Re-wording this section:

Sign up users normally with the chosen `Username` type and password. `"phone_number"` will be required in the `userAttributes` if SMS MFA is ON, or manually enabled for the user. Note that if the username attribute is the `phone_number`, then adding a separate attribute for phone number is not required during sign up.

You will need to pass `phone_number` as a user attribute to enable SMS MFA for your users during sign up. If your primary login mechanism on your Cognito resource is `phone_number` (without enabling `username`), then you do not need to pass it as an attribute.

import flutter6 from "/src/fragments/lib/auth/flutter/sms/sign_up.mdx";

<Fragments fragments={{flutter: flutter6}} />

import iosSignUp from "/src/fragments/lib/auth/ios/sms/sign_up.mdx";

<Fragments fragments={{ios: iosSignUp}} />

> Verification of user accounts is done via the `confirmSignUp` method with the OTP sent to their phone or email.

Reviewer 2: By default, you have to verify a user account after they sign up using the `confirmSignUp` API, which will send a one-time password to the user's phone number or email depending on your Amazon Cognito configuration.

import flutter7 from "/src/fragments/lib/auth/flutter/sms/confirm_sign_up.mdx";

<Fragments fragments={{flutter: flutter7}} />

import iosConfirmSignUp from "/src/fragments/lib/auth/ios/sms/confirm_sign_up.mdx";

<Fragments fragments={{ios: iosConfirmSignUp}} />

### Handling SMS MFA challenge during Sign In

> Sign in users normally with the chosen `Username` type and password.

Reviewer 2: After a user signs in, if they have MFA enabled for their account, a challenge will be returned that you would need to call the `confirmSignIn` API where the user provides their confirmation code sent to their phone number.

import flutter8 from "/src/fragments/lib/auth/flutter/sms/sign_in.mdx";

<Fragments fragments={{flutter: flutter8}} />

import iosSignInTOTP from "/src/fragments/lib/auth/ios/sms/sign_in.mdx";

<Fragments fragments={{ios: iosSignInTOTP}} />

If MFA is **ON** or enabled for the user, you must call `confirmSignIn` with the OTP sent to their phone.

import flutter9 from "/src/fragments/lib/auth/flutter/sms/confirm_sign_in.mdx";

<Fragments fragments={{flutter: flutter9}} />

import iosConfirmSignInTOTP from "/src/fragments/lib/auth/ios/sms/confirm_sign_in.mdx";

<Fragments fragments={{ios: iosConfirmSignInTOTP}} />

## Multi-factor authentication with TOTP

> You can use Time-based One-Time Password (TOTP) for multi-factor authentication (MFA) in your web or mobile applications. Amplify's Auth category includes support for handling TOTP setup and codes, offering an integrated solution for enhanced security for your users. Authenticator apps generate the temporary, one-time passwords. These apps, such as Google Authenticator, Microsoft Authenticator, have the TOTP algorithm built-in and work by using a shared secret key and the current time to generate a short-lived password.

Reviewer 2: Small nits:

You can use Time-based One-Time Password (TOTP) for multi-factor authentication (MFA) in your web or mobile applications. The Amplify Auth category includes support for handling TOTP setup and verification using authenticator apps, offering an integrated solution for enhanced security for your users. These apps, such as Google Authenticator, Microsoft Authenticator, have the TOTP algorithm built-in and work by using a shared secret key and the current time to generate a short-lived password.

<Callout>

<b>AWS Cognito Limitation</b>

The `fetchMFAPreference` API will only return enabled and preferred MFA methods, if the preferences were set using the `updateMFAPreference` API. If the preferences were not set using the `updateMFAPreference` API, the MFA preferences will return `nil` for enabled and preferred MFA methods. 

Therefore it is recommended to use the `updateMFAPreference` API to set the MFA preferences for the user, if the user decides to set up any MFA methods. An example use case would be that when a user successfully sets up TOTP using the `setupTOTP` and `verifyTOTPSetup` API, `updateMFAPreference` should be invoked to set the MFA preferences for the user i.e. marking TOTP as enabled.

</Callout>

> ### Handling TOTP Setup during sign in

Reviewer 2: change the title to "Setting up TOTP for a user"

A `signIn` call will return `continueSignInWithTOTPSetup` as a next step when the following set of conditions are met:

- MFA is marked as **Required** in Cognito User Pool.
- TOTP is enabled in the Cognito User Pool
- User does not have any other MFA method set up.

> The `continueSignInWithTOTPSetup` step signifies that the user must set up TOTP before they can sign in. The step returns an associated value of type `TOTPSetupDetails` which must be used to configure an Authenticator app like Microsoft Authenticator or Google Authenticator. `TOTPSetupDetails` provides a helper method called `getSetupURI` which generates a URI that can be used, for example, in a button to open the user's installed Authenticator app. For more advanced use cases, `TOTPSetupDetails` also contains a `sharedSecret` which can be used to either generate a QR code or be manually entered into an Authenticator app. 

Reviewer 2: authenticator app is lower case

> Once the Authenticator app is set up, the user can generate a TOTP code and provide it to the library to complete the sign in process.

Reviewer 2: authenticator is lower case here.

**`verifyTOTPSetup` API**

import iosVerifyTOTPSetup from "/src/fragments/lib/auth/ios/totp/verify_totp_setup.mdx";

<Fragments fragments={{ios: iosVerifyTOTPSetup}} />

### Recovering from a lost TOTP device

<Callout warning>

If a user loses access to their TOTP device, they would need to contact an administrator to help get access to their account. Based on the Cognito User Pool configuration, the administrator can use the [AdminSetUserMFAPreference](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_AdminSetUserMFAPreference.html) to either change the MFA preference to a different MFA method or to disable MFA for the user.

</Callout>

In a scenario where MFA is marked as **Required** in Cognito User Pool and another MFA method is not set up, the administrator would need to first initiate an [AdminUpdateUserAttributes](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_AdminUpdateUserAttributes.html) call and update the user’s phone number attribute. Once this is complete, the administrator can continue changing the MFA preference to SMS as suggested above.

## Setting a user's preferred MFA option

<Callout>

> The APIs for fetching and updating MFA preferences are only available through AWS Cognito Auth plugin. The examples usage of the API's using the plugin is provided below. 

Reviewer 2: why is this important to call out here? Let us remove it. If this is because of the Auth0 plugin, I don't think this callout should be here.

Reviewer: It's called out because they're not available on the category, only the plugin.

Reviewer 2: My question here is about why do we need to have that be a callout? Can we just not provide the code snippet and describe the behavior and move on? Why would a developer care?

</Callout>

### Fetch the current user's MFA preferences

Invoke the following API to get the MFA preference for the current user.

import iosFetchMFAPreference from "/src/fragments/lib/auth/ios/mfa_preference/10_fetch_mfa_preference.mdx";

<Fragments fragments={{ios: iosFetchMFAPreference}} />

### Update the current user's MFA preferences

Invoke the following API to update the MFA preference for the current user.

<Callout warning>

Only one MFA method can be marked as preferred at a time. If the user has multiple MFA methods enabled and tries to mark more than one MFA method as preferred, the API will throw an error.

</Callout>

import iosUpdateMFAPreferences from "/src/fragments/lib/auth/ios/mfa_preference/20_update_mfa_preference.mdx";

<Fragments fragments={{ios: iosUpdateMFAPreferences}} />

> If multiple MFA methods are enabled for the user, the `signIn` API will return `continueSignInWithMFASelection` as the next step in the auth flow. During this scenario, the user should be prompted to select the MFA method they want to use to sign in.

Reviewer 2: we need a code-snippet for this. How would this work? Also - I feel like this should probably have its own header

<Callout>

> Refer to the [Continue signin with MFA Selection](/lib/auth/signin_next_steps#continue-signin-with-mfa-selection) section for more details.

Reviewer 2: The details for this step should be here - rather than in a seperate page. We should add the code snippet and remove that mention in the other page.

Reviewer 2: the callout would then be removed

Reviewer: Well, that page is specifically about enumerating all the steps which can come back from `confirmSignIn`. If we want it on this page, it probably makes sense to have it on both. Otherwise, the other page would be incomplete as a reference.

Reviewer 2: got it - let us have it on both. I am eventually planning on redistributing the content in that page.

<Callout warning>

When using Amplify CLI to set up backend resources, the following options are only available when starting a new project (via `amplify add auth`). You will not have access to these settings after creation (via `amplify update`).

- Required MFA (i.e. Setting MFA to "ON")

</Callout>

<Callout>
Note: If you create or update an SMS MFA configuration for your Cognito user pool, the Cognito service will send a test SMS message to an internal number in order to verify your configuration. You will be charged for these test messages by Amazon SNS. 

Refer to [SMS sandbox](/lib/auth/sms_flows#sandbox-mode).

For information about Amazon SNS pricing, see [Worldwide SMS Pricing](https://aws.amazon.com/sns/sms-pricing/). 
</Callout>

<BlockSwitcher>

<Block name="New Project">

Run `amplify add auth` to create a new Cognito Auth resource, and follow the prompts below depending on how you want to integrate MFA into your flow.

Turning MFA "ON" will make it required for all users, while "Optional" will make it available to enable on a per-user basis.

import all2 from "/src/fragments/lib/auth/common/mfa/add_mfa.mdx";

<Fragments fragments={{all: all2}} />

</Block>

> <Block name="Existing Auth">

Reviewer 2: can we change this title to Existing Project

<Fragments fragments={{ios: iosUpdateMFAPreferences}} />

### Enabling TOTP after a user is signed in

TOTP MFA can be set up after a user has signed in. This can be done when the following conditions are met:

- MFA is marked as **Optional** or **Required** in the Cognito User Pool
- TOTP is marked as an enabled MFA method in Cognito user pool

TOTP can be set up by calling the `setUpTOTP` and `verifyTOTPSetup` APIs in the `Auth` category.

Invoke the `setUpTOTP` API to generate a `TOTPSetupDetails` object which should be used to configure an Authenticator app like Microsoft Authenticator or Google Authenticator. `TOTPSetupDetails` provides a helper method called `getSetupURI` which generates a URI that can be used, for example, in a button to open the user's installed Authenticator app. For more advanced use cases, `TOTPSetupDetails` also contains a `sharedSecret` which can be used to either generate a QR code or be manually entered into an Authenticator app.

that contains the `sharedSecret` which will be used to either to generate a QR code or can be manually entered into an Authenticator app.

> **`setUpTOTP` API**

Reviewer 2: can we remove this please

import iosSetUpTOTP from "/src/fragments/lib/auth/ios/totp/set_up_totp.mdx";

<Fragments fragments={{ios: iosSetUpTOTP}} />

Once the Authenticator app is set up, the user must generate a TOTP code and provide it to the library. Pass the code to `verifyTOTPSetup` to complete the TOTP setup process.

> **`verifyTOTPSetup` API**

Reviewer 2: can we remove this - we describe it in the sentence before

> ## SMS flows

Reviewer 2: for the intro of this section - it should just state that in order to secure your app, you can add MFA. And that when you add MFA, you have two options to do that, whether through SMS or Time-based one-time passwords. All this info about how phone numbers are used are unnecessary.

Author: Done.

For information about Amazon SNS pricing, see [Worldwide SMS Pricing](https://aws.amazon.com/sns/sms-pricing/).
</Callout>

There are a few ways to integrate phone numbers into an Amplify project's sign-in and verification process.
- **As a username**\*: Users login with a username and password where their phone number acts as the username.
- **As a verification method**: Users login by any means, but must verify their account with an OTP (one time password) sent to their phone.
- **MFA (Multi-Factor Authentication)**: Users must verify every login with an OTP sent to their phone.

\**Note: This is different from using a phone number [alias](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html#user-pool-settings-aliases), which is currently unsupported by the Amplify CLI.*

These methods may be combined with each other or used independently but they all require the same prerequisites for sending SMS messages via Amazon SNS, the notification service used by Amplify.

> ### Prerequisites

Reviewer 2: for the sections on the right nav - we should have them be in the order:

- High level intro (doesn't need a header - and it should just be the name on the sidenav which is "Multi-factor Authentication".
- Set up backend resources (this section should detail out the experience for a developer using the CLI to add SMS or TOTP).
- Multi-factor authentication with SMS
- Multi-fcator authentication with TOTP

All the info here about the sign up and using phone numbers for registration should be removed. This guide should be all about MFA.

Author: Done.

**`verifyTOTPSetup` API**

import iosVerifyTOTPSetup from "/src/fragments/lib/auth/ios/totp/verify_totp_setup.mdx";

<Fragments fragments={{ios: iosVerifyTOTPSetup}} />

### Recovering from a lost TOTP device

If a user loses access to their TOTP device, they would need to contact an administrator to help get access to their account. Based on the Cognito User Pool configuration, the administrator can use the [AdminSetUserMFAPreference](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_AdminSetUserMFAPreference.html) to either change the MFA preference to a different MFA method or to disable MFA for the user.

<Callout warning>

In a scenario where MFA is marked as **Required** in Cognito User Pool and another MFA method is not set up, the administrator would need to first initiate an [AdminUpdateUserAttributes](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_AdminUpdateUserAttributes.html) call and update the user’s phone number attribute. Once this is complete, the administrator can continue changing the MFA preference to SMS as suggested above.

</Callout>

> ## MFA Preferences

Reviewer 2: please change this title to "Setting up user preferred MFA option"

Author: Done.

Invoke the following API to get the MFA preference for the current user.

import iosFetchMFAPreference from "/src/fragments/lib/auth/ios/mfa_preference/10_fetch_mfa_preference.mdx";

<Fragments fragments={{ios: iosFetchMFAPreference}} />

### Update the current user's MFA preferences

Invoke the following API to update the MFA preference for the current user.

<Callout warning>

Only one MFA method can be marked as preferred at a time. If the user has multiple MFA methods enabled and tries to mark more than one MFA method as preferred, the API will throw an error.

</Callout>

import iosUpdateMFAPreferences from "/src/fragments/lib/auth/ios/mfa_preference/20_update_mfa_preference.mdx";

<Fragments fragments={{ios: iosUpdateMFAPreferences}} />

If multiple MFA methods are enabled for the user, the `signIn` API will return `continueSignInWithMFASelection` as the next step in the auth flow. During this scenario, the user should be prompted to select the MFA method they want to use to sign in.

Refer to the [Continue signin with MFA Selection](/lib/auth/signin_next_steps#continue-signin-with-mfa-selection) section for more details.

<Callout>

> <b>AWS Cognito Limitation</b>

Reviewer 2: For this callout - I think this should just be a part of the SMS MFA and TOTP setup flows - that you have to call the set MFA preference after set up so that you can then retrieve this information when you call the getMFAPreferences API

Author: Done. Moved call outs to both SMS and TOTP sections.

Author: @dnys1 This was the reason I moved the callouts in the other sections rather than preferences.
